File: Script/MergeSort/MergeSort.py
def merge_sort(arr):
    if len(arr) <= 1:
        return arr

    # 1. DIVIDI
    mid = len(arr) // 2
    left = arr[:mid]
    right = arr[mid:]

    # 2. RICORRI
    left = merge_sort(left)
    right = merge_sort(right)

    # 3. CONQUISTA (MERGE)
    return merge(left, right)

def merge(left, right):
    result = []
    i = j = 0

    while i < len(left) and j < len(right):
        if left[i] < right[j]:
            result.append(left[i])
            i += 1
        else:
            result.append(right[j])
            j += 1

    # Aggiungi i rimanenti
    result.extend(left[i:])
    result.extend(right[j:])
    return result

# --- Gestione Input/Output ---
def main():
    try:
        with open('input.txt', 'r') as f:
            # Leggi i numeri, rimuovi spazi bianchi e li converte in interi
            data_str = f.read().split()
            arr = [int(x) for x in data_str if x.isdigit()]
            
            print("Array non ordinato:", arr)
            sorted_arr = merge_sort(arr)
            print("Array ordinato:", sorted_arr)

    except FileNotFoundError:
        print("Errore: il file 'input.txt' non è stato trovato.")
    except Exception as e:
        print(f"Si è verificato un errore: {e}")

if __name__ == "__main__":
    main()

File: Script/MergeSort/MergeSort.js
const fs = require('fs');


function mergeSort(arr) {
    if (arr.length <= 1) {
        return arr;
    }

    // 1. DIVIDI
    const mid = Math.floor(arr.length / 2);
    const left = arr.slice(0, mid);
    const right = arr.slice(mid);

    // 2. RICORRI e 3. CONQUISTA (MERGE)
    return merge(mergeSort(left), mergeSort(right));
}

function merge(left, right) {
    let result = [];
    let i = 0;
    let j = 0;

    while (i < left.length && j < right.length) {
        if (left[i] < right[j]) {
            result.push(left[i]);
            i++;
        } else {
            result.push(right[j]);
            j++;
        }
    }

    // Aggiungi i rimanenti
    return result.concat(left.slice(i)).concat(right.slice(j));
}

// --- Gestione Input/Output (Node.js) ---
try {
    // Legge il contenuto del file come stringa
    const data = fs.readFileSync('Script/MergeSort/input.txt', 'utf8');
    
    // Filtra e converte in numeri interi
    const arr = data
        .trim() // Rimuovi spazi bianchi all'inizio/fine
        .split(/\s+/) // Dividi per qualsiasi spazio bianco (spazi, newline)
        .map(Number)
        .filter(n => !isNaN(n)); // Assicura che sia un numero valido

    console.log("Array non ordinato:", arr);
    const sortedArr = mergeSort(arr);
    console.log("Array ordinato:", sortedArr);

} catch (err) {
    if (err.code === 'ENOENT') {
        console.error("Errore: il file 'input.txt' non è stato trovato.");
    } else {
        console.error("Si è verificato un errore:", err.message);
    }
}